File: 235.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Node n = func(root, p, q);
        return n.node;
    }
    private Node func(TreeNode root, TreeNode p, TreeNode q){
        int cnt = 0;
        if (root == p || root == q){
            cnt = 1;
        }
        Node n;
        if (root.left != null){
            n = func(root.left, p, q);
            if (n.node != null){
                return n;
            }
            cnt += n.cnt;
        }
        if (root.right != null){
            n = func(root.right, p, q);
            if (n.node != null){
                return n;
            }
            cnt += n.cnt;
        }
        if (cnt == 2){
            return new Node(root, 2);
        }
        return new Node(null, cnt);
    }
}

class Node{
    public TreeNode node;
    public int cnt;
    public Node(TreeNode node, int cnt){
        this.cnt = cnt;
        this.node = node;
    }
}

File: 235.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def func(root):
            cnt = 0
            if root == p or root == q:
                cnt += 1
            if root.left:
                a = func(root.left)
                if type(a) is int:
                    cnt += a
                else:
                    return a
            if root.right:
                b = func(root.right)
                if type(b) is int:
                    cnt += b
                else:
                    return b
            if cnt == 2:
                return root
            return cnt
        return func(root)